File: gps_tracker/src/route_logger.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <nav_msgs/msg/path.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <cmath>

class RouteLogger : public rclcpp::Node
{
public:
    RouteLogger() : Node("route_logger")
    {
        subscription_ = this->create_subscription<sensor_msgs::msg::NavSatFix>(
            "/gps/fix",
            10,
            std::bind(&RouteLogger::gps_callback,this,std::placeholders::_1)
        );

        publisher_ = this->create_publisher<nav_msgs::msg::Path>("/gps/path",10);

        path_msg_.header.frame_id = "map";
    }

private:

    void gps_callback(const sensor_msgs::msg::NavSatFix::SharedPtr msg)
    {
        if(!origin_set_)
        {
            origin_lat_ = msg->latitude;
            origin_lon_ = msg->longitude;
            origin_set_ = true;
        }

        double x,y;
        latlon_to_xy(msg->latitude,msg->longitude,x,y);

        geometry_msgs::msg::PoseStamped pose;

        pose.header.stamp = this->now();
        pose.header.frame_id = "map";

        pose.pose.position.x = x;
        pose.pose.position.y = y;
        pose.pose.position.z = 0.0;

        pose.pose.orientation.w = 1.0;

        path_msg_.header.stamp = this->now();
        path_msg_.poses.push_back(pose);

        publisher_->publish(path_msg_);
    }

    void latlon_to_xy(double lat,double lon,double &x,double &y)
    {
        constexpr double R = 6378137.0;

        double dlat = (lat-origin_lat_) * M_PI/180.0;
        double dlon = (lon-origin_lon_) * M_PI/180.0;
        double mean_lat = ((lat+origin_lat_)/2.0) * M_PI/180.0;

        y = R * dlat;
        x = R * cos(mean_lat) * dlon;
    }

    rclcpp::Subscription<sensor_msgs::msg::NavSatFix>::SharedPtr subscription_;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr publisher_;

    nav_msgs::msg::Path path_msg_;

    bool origin_set_ = false;
    double origin_lat_;
    double origin_lon_;
};

int main(int argc,char * argv[])
{
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<RouteLogger>());
    rclcpp::shutdown();
    return 0;
}

File: autonomy_sensors/autonomy_sensors/route_logger.py
import math

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import NavSatFix
from nav_msgs.msg import Path
from geometry_msgs.msg import PoseStamped


class RouteLogger(Node):

    def __init__(self):
        super().__init__('route_logger')

        self.subscription_ = self.create_subscription(
            NavSatFix,
            '/gps/fix',
            self.gps_callback,
            10
        )
        self.publisher_ = self.create_publisher(Path, '/gps/path', 10)

        self.path_msg_ = Path()
        self.path_msg_.header.frame_id = 'map'

        self.origin_set_ = False
        self.origin_lat_ = 0.0
        self.origin_lon_ = 0.0

    def gps_callback(self, msg: NavSatFix):
        if not self.origin_set_:
            self.origin_lat_ = msg.latitude
            self.origin_lon_ = msg.longitude
            self.origin_set_ = True

        x, y = self.latlon_to_xy(msg.latitude, msg.longitude)

        pose = PoseStamped()
        pose.header.stamp = self.get_clock().now().to_msg()
        pose.header.frame_id = 'map'
        pose.pose.position.x = x
        pose.pose.position.y = y
        pose.pose.position.z = 0.0
        pose.pose.orientation.w = 1.0

        self.path_msg_.header.stamp = self.get_clock().now().to_msg()
        self.path_msg_.poses.append(pose)

        self.publisher_.publish(self.path_msg_)

    def latlon_to_xy(self, lat: float, lon: float):
        R = 6378137.0

        dlat = (lat - self.origin_lat_) * math.pi / 180.0
        dlon = (lon - self.origin_lon_) * math.pi / 180.0
        mean_lat = ((lat + self.origin_lat_) / 2.0) * math.pi / 180.0

        y = R * dlat
        x = R * math.cos(mean_lat) * dlon
        return x, y


def main(args=None):
    rclpy.init(args=args)
    node = RouteLogger()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()